### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by atoledo-ec**
## Friendship v2

In this branch I implement friendships with 2 rows for mutual friendship. Changing the associations and modifying the user methods used before.

### CODE REVIEW

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've really done well implementing the milestone requirements. You also did well to add unit and integration test suites.
However, there are some changes required before this milestone can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] I could not test your implemented feature.
- Here's the error I'm having whenever I try to access your project:



- Although I do not understand why you're having the RSpec configuration in the `user_helper.rb` file of your `helpers` directory, Kindly check this file to fix this error.

**Having checked your codebase, here are some hints about implementing mutual friendship in the most efficient way.**

- [x] When friendship is confirmed, you can create an additional row by adding this code snippet into your friendship model, like this:

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

- [x] Thanks to the snippet above, you can find user friends only by user_id in the friendships table:
```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships

```

- [x] Also, User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
 
```

- [x] User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data, it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

- [x] In order to get all posts written by user and their friends, you can use simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

**[ COMMENTS ]**
- Please note that the explanations above concerning implementing mutual friendship are to enable you to see an efficient implementation of mutual friendship.

- In my opinion, it does not mean that you should completely refactor yours to use the method stated above. However, you are advised to see how you can use it to improve your own version of the implementation.


* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Comment by atoledo-ec** [comment]
In order to test the helper methods I read that recommends to put the configuration inside the `user_helper.rb`

**Comment by teekaytech** [comment]
> In order to test the helper methods I read that recommends to put the configuration inside the `user_helper.rb`
> 

Okay. If that's the case, then you should find a way to make it work without affecting the project flow. All the best.

**Comment by atoledo-ec** [comment]
Now I can access the project with `rails server` but the test for feature are not working, if you have any advise please let me know

**Comment by atoledo-ec** [comment]
Also I put the configuration in the `rails_helper.rb` instead on the `user_helper.rb`

File: app/models/user.rb
Comment (tirthajyoti-ghosh) [inline app/models/user.rb]: These two methods can be converted to scopes. 

Right now, you are using Ruby methods to process pending friendships and friend requests. But if you have used ActiveRecord methods, the processing would have been a lot faster as the database was created to store and query data it is always more efficient than Ruby.

- [x] Please convert these methods to scopes.

File: app/views/users/_pending.html.erb
Comment (tirthajyoti-ghosh) [inline app/views/users/_pending.html.erb]: Please remove all logic from your views. Move them to helpers.

**IMPORTANT!** I am only pointing this out here. Please go through your views and refactor the code as necessary.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## Impressive progress so far 💪! Thanks for making all the changes 👏.

The way you have used ActiveRecord methods is truly impressive especially the `friend_post` scope in post.rb. I have seen very few people handle ActiveRecords that well. Well done 👏 !

### However, we need to work on some changes before we can merge.

After rejecting a friendship, I can still see the "accept" and "reject" button beside the user on the "All Users" page and on the user profile page. You can use server logs to see what's happening. Please remove the buttons after the logged-in user has rejected friendship.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Comment by atoledo-ec** [comment]
I tried to reject a friendship request and then the buttons for accept and reject disappear, please check this again.

**Comment by tirthajyoti-ghosh** [comment]
Try again after creating 2 or 3 friendships. You will see the buttons still
remains.

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

A great effort implementing the suggested changes 👍 
However, we still need to make some changes in order to get this project approved 💪 
Kindly check my comments below 👇  and resubmit your project once they are done ⏩ 
Good luck! 😃 
## Mandatory changes
- The project has a bug due to the associations and foreign keys you have implemented. Try to move your scopes into the users model, and double-check the foreign keys in your friendships models

**Comment by atoledo-ec** [comment]
Please can you explain to me more about the bug you are talking about

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @alejandrotoledoweb :+1:
#### Good job so far. However, the database migration process outputted an error when I was trying to run `rails db:migrate`. Check the image and the comment below :point_down:

#### Migration Error:
- _The migration process output this error: "rails aborted!
ActiveModel::UnknownAttributeError: unknown attribute 'confirmed' for Friendship."_
Screenshot from 2020-12-17 14-14-36

#### _Review by Temitayo Adeleke_

File: app/views/users/_pending.html.erb
Comment (teekaytech) [inline app/views/users/_pending.html.erb]: Again, this is a block of logic inside your view. 
Kindly move this to a helper file.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've done amazingly well so far. All previously required changes have been implemented. Everything seems to be working.
Two major issues are preventing me from approving your project. Let's look into it.__

**[ REQUIRED CHANGES ]**

- [ ] In the `All Users` page, there is a strange output printed below the users. Check the image below for details.



- [ ] Secondly, some of your test cases are failing. Check the image below to see the affected test cases.





* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello Team  ,  you have done a brilliant job with this project. :clap:
#### You can merge !!!

**Pull request merged**
